web/modules/custom/nfafmis/nfafmis.module

```diff
@@ -798,3 +759,199 @@ function nfafmis_form_ajax_validate(array &$form, FormStateInterface $form_state
   $form_state->setResponse($response);
   return $response;
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function nfafmis_form_user_form_alter(array &$form, FormStateInterface $form_state) {
+  // Show the Management unit/Range field for Range users only.
+  $form['management_unit']['#states'] = [
+    'visible' => [
+      [':input[name="roles[range_user]"]' => ['checked' => TRUE]],
+      [':input[name="roles[range_power_user]"]' => ['checked' => TRUE]],
+    ],
+  ];
+  $form['management_unit']['widget']['0']['target_id']['#states'] = [
+    'required' => [
+      [':input[name="roles[range_user]"]' => ['checked' => TRUE]],
+      [':input[name="roles[range_power_user]"]' => ['checked' => TRUE]],
+    ],
+  ];
+
+  // Only users with account admin permissions can edit the management unit.
+  if (!\Drupal::currentUser()->hasPermission('administer users')) {
+    $form['management_unit']['#disabled'] = 'disabled';
+  }
+}
+
+/**
+ * Implements hook_node_access_records().
+ * Set the rules that determine who has access to this $node.
+ */
+function nfafmis_node_access_records(NodeInterface $node) {
+  $grants = [];
+
+  // Content types that will be restricted according to the grants of the user
+  // accessing the node.
+  $bundles = ['farmer_details', 'offer_license', 'sub_area'];
+  if (!in_array($node->bundle(), $bundles) || !$node->isPublished()) {
+    // Access to other content types will be determined by standard permissions.
+    return $grants;
+  }
+
+  /** @var \Drupal\Core\Config\Entity\ConfigEntityStorage $entity_storage */
+  $node_storage = \Drupal::entityTypeManager()->getStorage('node');
+  $range = NULL;
+
+  if ($node->bundle() == 'farmer_details') {
+    // Retrieve the Areas that the farmer has allocations in, and from there the
+    // CFRs and Ranges (Management Unit) that the area is in.
+    // Load Sub area nodes ids.
+    $nids = \Drupal::entityQuery('node')
+      ->condition('type', 'offer_license')
+      ->condition('field_farmer_name_ref', $node->id())
+      ->accessCheck(FALSE)
```

Comment:
I think we need accessCheck. We're setting permission to access the farmer node, which can only be determined by looking at the ranges that the farmer's areas are in. If for some reason the user saving the farmer node doesn't have permission to access the area node we still need to know what range the area is in so that we can grant access to the farmer node. 
You're right, we don't want the unpublished nodes. I'll filter them out